File: agent_arena/agents/loader.py
"""Load agent configurations from YAML files."""

from pathlib import Path
from typing import List
import yaml
import logging

from ..core.types import AgentConfig
from .agent import Agent

logger = logging.getLogger(__name__)


def load_agent_config(path: Path | str) -> AgentConfig:
    """Load an agent configuration from a YAML file."""
    path = Path(path)

    if not path.exists():
        raise FileNotFoundError(f"Agent config not found: {path}")

    with open(path, "r") as f:
        config = yaml.safe_load(f)

    # Set ID from filename if not specified
    if "id" not in config:
        config["id"] = path.stem

    return config


def load_agents_from_directory(directory: Path | str) -> List[Agent]:
    """Load all agent configurations from a directory."""
    directory = Path(directory)
    agents = []

    if not directory.exists():
        logger.warning(f"Agent directory not found: {directory}")
        return agents

    for yaml_file in directory.glob("*.yaml"):
        try:
            config = load_agent_config(yaml_file)
            agent = Agent.from_config(config)
            agents.append(agent)
            logger.info(f"Loaded agent: {agent.name} from {yaml_file}")
        except Exception as e:
            logger.error(f"Failed to load agent from {yaml_file}: {e}")

    return agents


def create_agent_from_dict(data: dict) -> Agent:
    """Create an agent from a dictionary (e.g., from API request)."""
    return Agent.from_config(data)

File: src/agents/loader.ts
/**
 * Load agent configurations from YAML files.
 */

import { readFileSync, readdirSync, existsSync } from 'fs';
import { join, parse } from 'path';
import yaml from 'js-yaml';
import { AgentConfig } from '../core/types.js';
import { Agent } from './agent.js';

/**
 * Load an agent configuration from a YAML file.
 */
export function loadAgentConfig(filePath: string): AgentConfig {
  if (!existsSync(filePath)) {
    throw new Error(`Agent config not found: ${filePath}`);
  }

  const content = readFileSync(filePath, 'utf-8');
  const config = yaml.load(content) as AgentConfig;

  // Set ID from filename if not specified
  if (!config.id) {
    config.id = parse(filePath).name;
  }

  return config;
}

/**
 * Load all agent configurations from a directory.
 */
export function loadAgentsFromDirectory(directory: string): Agent[] {
  const agents: Agent[] = [];

  if (!existsSync(directory)) {
    console.warn(`Agent directory not found: ${directory}`);
    return agents;
  }

  const files = readdirSync(directory);

  for (const file of files) {
    if (!file.endsWith('.yaml') && !file.endsWith('.yml')) {
      continue;
    }

    const filePath = join(directory, file);
    try {
      const config = loadAgentConfig(filePath);
      const agent = Agent.fromConfig(config);
      agents.push(agent);
      console.log(`Loaded agent: ${agent.name} from ${filePath}`);
    } catch (error) {
      console.error(`Failed to load agent from ${filePath}:`, error);
    }
  }

  return agents;
}

/**
 * Create an agent from a dictionary (e.g., from API request).
 */
export function createAgentFromDict(data: AgentConfig): Agent {
  return Agent.fromConfig(data);
}
